File: server/src/utils/gpg.js
const { spawnSync } = require("child_process");
const commandExistsSync = require("command-exists").sync;

/**
 * checks if gpg is installed in path
 *
 * @returns {boolean} boolean result if gpg is installed
 */
const _gpgExists = () => {
    return commandExistsSync("gpg");
};

/**
 * signs data and returns output as bytes
 *
 * @param {string} keyId - id of key
 * @param {bytes} data - data to sign
 * @returns {bytes} signed data
 */
const sign = (keyId, data) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", ["--default-key", keyId, "--output", "-", "--sign"], {
        input: data,
    });

    if (gpg.status !== 0) throw new Error(gpg.stderr);

    return gpg.stdout;
};

/**
 * verify and returns data
 *
 * @param {string} keyId - key to use
 * @param {bytes} data - data to verify
 * @returns {bytes} signed content
 */
const verify = (keyId, data) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", ["--decrypt"], {
        input: data,
    });

    if (gpg.status !== 0) throw new Error(gpg.stderr);

    if (!gpg.stderr.includes(` ${keyId}\n`) && !gpg.stderr.includes(` ${keyId}\r`))
        throw new Error("Invalid Key ID");

    return gpg.stdout;
};

/**
 * encrypt data and returns output as bytes
 *
 * @param {string} keyId - key to use
 * @param {bytes} data - data to encrypt
 * @returns {bytes} encrypted data
 */
const encrypt = (keyId, data) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync(
        "gpg",
        ["--recipient", keyId, "--trust-model", "always", "--output", "-", "--encrypt"],
        {
            input: data,
        }
    );

    if (gpg.stderr.length !== 0) throw new Error(gpg.stderr.toString());

    return gpg.stdout;
};

/**
 * decrypt data and returns output as bytes
 *
 * @param {bytes} data - data to decrypt
 * @returns {bytes} decrypted data
 */
const decrypt = (data) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", ["--decrypt"], {
        input: data,
    });

    if (gpg.status !== 0) throw new Error(gpg.stderr.toString());

    return gpg.stdout;
};

/**
 * imports key and returns key ID
 *
 * @param {bytes} data - public key in bytes
 * @returns {string} key id
 */
const importKey = (data) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", ["--import-options", "import-show", "--import"], {
        input: data,
    });

    if (gpg.status !== 0) throw new Error(gpg.stderr.toString());

    return gpg.stdout.toString().match(/.+ \[expires: .+\]\s+(.+)/)[1];
};

/**
 * returns public key as bytes
 *
 * @param {string} keyId - key to export
 * @returns {bytes} exported key
 */
const exportKey = (keyId) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", ["--export", keyId]);

    if (gpg.stderr.length !== 0) throw new Error(gpg.stderr.toString());

    return gpg.stdout;
};

/**
 * generates key pair
 *
 * @param {string} keyEmail - key email
 * @param {string} password - password to encrypt key
 * @returns {boolean} boolean of successful generation
 */
const genKey = (keyEmail, password) => {
    if (!_gpgExists()) throw new Error("GPG command does not exist");

    const gpg = spawnSync("gpg", [
        "--batch",
        "--passphrase",
        password,
        "--quick-gen-key",
        `OAK <${keyEmail}>`,
        "default",
        "default",
    ]);

    if (gpg.status !== 0) throw new Error(gpg.stderr.toString());
    return gpg.stderr.toString().match(/\/([A-F0-9]{40})\./)[1];
};

module.exports = {
    sign: sign,
    verify: verify,
    encrypt: encrypt,
    decrypt: decrypt,
    importKey: importKey,
    exportKey: exportKey,
    genKey: genKey,
};

File: client/oak/gpg.py
#!/usr/bin/env python3
import subprocess


def _gpg_exists() -> bool:
    """
    Checks if gpg is installed in path

    :return: Boolean result if gpg is installed
    """
    from shutil import which
    return which("gpg") is not None


def sign(key_id: str, data: bytes, options: list = []) -> bytes:
    """
    Signs data and returns output as bytes

    :param key_id: Key To Use
    :param data: Data To Sign
    :param options: Additional Options
    :return: Signed Data
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(["gpg", "--default-key", key_id, "--output", "-"] + options +
                               ["--sign"],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    stdout, stderr = process.communicate(input=data)

    if process.wait() != 0:
        raise Exception(stderr)

    return stdout


def verify(key_id: str, data: bytes, options: list = []) -> bytes:
    """
    Verify and returns data

    :param key_id: Key To Use
    :param data: Data To Verify
    :param options: Additional Options
    :return: Signed Content
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(["gpg"] + options + ["--decrypt"],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    stdout, stderr = process.communicate(input=data)

    if process.wait() != 0:
        raise Exception(stderr)

    if f" {key_id}\n".encode() not in stderr:
        raise Exception("Invalid Key ID")

    return stdout


def encrypt(key_id: str, data: bytes, options: list = []) -> bytes:
    """
    Encrypt data and returns output as bytes

    :param key_id: Key To Use
    :param data: Data To Encrypt
    :param options: Additional Options
    :return: Encrypted Data
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(
        ["gpg", "--recipient", key_id, "--trust-model", "always", "--output", "-"] + options +
        ["--encrypt"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE)
    stdout, stderr = process.communicate(input=data)

    if stderr:
        raise Exception(stderr)

    return stdout


def decrypt(data: bytes, options: list = []) -> bytes:
    """
    Decrypt data and returns output as bytes

    :param key_id: Key To Use
    :param data: Data To Decrypt
    :param options: Additional Options
    :return: Decrypted Data
    """
    if not _gpg_exists():
        raise Exception("GPG does not exist")

    process = subprocess.Popen(["gpg"] + options + ["--decrypt"],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    stdout, stderr = process.communicate(input=data)

    if process.wait() != 0:
        raise Exception(stderr)

    return stdout


def import_key(data: bytes, options: list = []) -> bool:
    """
    Imports key and returns key ID

    :param data: Public Key In Bytes
    :param options: Additional Options
    :return: Key ID
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(["gpg"] + options + ["--import-options", "import-show", "--import"],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    stdout, stderr = process.communicate(input=data)

    if process.wait() != 0:
        raise Exception(stderr)

    from re import findall
    return findall('''.+ \[expires: .+\]\s+(.+)''', stdout.decode())[0]


def export_key(key_id: str, options: list = []) -> bytes:
    """
    Returns public key as bytes

    :param key_id: Key To Export
    :param options: Additional Options
    :return: Exported Key
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(["gpg"] + options + ["--export", key_id],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()

    if stderr:
        raise Exception(stderr)

    return stdout


def gen_key(key_email: str, password: str, options: list = []) -> str:
    """
    Generates key pair

    :param key_email: Key Email
    :param password: Password To Encrypt Key
    :param options: Additional Options
    :return: Key ID of generated key
    """
    if not _gpg_exists():
        raise Exception("GPG command does not exist")

    process = subprocess.Popen(["gpg"] + options + [
        "--batch", "--passphrase", password, "--quick-gen-key", f"OAK <{key_email}>", "default",
        "default"
    ],
                               stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)

    _, stderr = process.communicate()

    if process.wait() != 0:
        raise Exception(stderr)

    from re import findall
    return findall('''/([A-F0-9]{40})\.''', stderr.decode())[0]
